### PR TITLE
[PM-5974] introduce ForwarderGeneratorStrategy

### PR DESCRIPTION
**Pull request by audrey-jensen**
## Type of change

```
- [ ] Bug fix
- [ ] New feature development
- [X] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Introduce `ForwarderGeneratorStrategy`, a `GeneratorStrategy` that integrates email forwarders into the `DefaultGeneratorService`.

Note: this PR does not require QA because it is not yet used by any UI logic

## Code changes

- **libs/common/src/tools/generator/key-definition.ts:** add forwarder key definitions
- **libs/common/src/tools/generator/username/forwarder-generator-strategy.ts:** common abstraction for all email forwarders
- **libs/common/src/tools/generator/username/options/forwarder-options.ts:** Remove `Forwarder` because it's replaced by the strategy. Introduce `RequestOptions` to integrate forwarders with the generator strategy pattern.
- **libs/common/src/tools/generator/username/options/constants.ts:** default `website` to `null`.

Implemented `ForwarderGenerationStrategy` on all forwarders:
- **libs/common/src/tools/generator/username/forwarders/addy-io.ts** 
- **libs/common/src/tools/generator/username/forwarders/duck-duck-go.ts** 
- **libs/common/src/tools/generator/username/forwarders/fastmail.ts** 
- **libs/common/src/tools/generator/username/forwarders/firefox-relay.ts** 
- **libs/common/src/tools/generator/username/forwarders/forward-email.ts** 
- **libs/common/src/tools/generator/username/forwarders/simple-login.ts** 

Unit tests:
- **libs/common/src/tools/generator/key-definition.spec.ts** 
- **libs/common/src/tools/generator/username/forwarder-generator-strategy.spec.ts** 
- **libs/common/src/tools/generator/username/forwarders/addy-io.spec.ts** 
- **libs/common/src/tools/generator/username/forwarders/duck-duck-go.spec.ts** 
- **libs/common/src/tools/generator/username/forwarders/fastmail.spec.ts** 
- **libs/common/src/tools/generator/username/forwarders/firefox-relay.spec.ts** 
- **libs/common/src/tools/generator/username/forwarders/forward-email.spec.ts** 
- **libs/common/src/tools/generator/username/forwarders/simple-login.spec.ts** 
- **libs/common/src/tools/generator/username/options/utilities.spec.ts**

### CODE REVIEW

File: libs/common/src/tools/generator/username/forwarders/forward-email.ts [hunk 1]
```diff
@@ -1,25 +1,39 @@
...
     private i18nService: I18nService,
-  ) {}
+    encryptService: EncryptService,
+    keyService: CryptoService,
+    stateProvider: StateProvider,
+  ) {
+    super(encryptService, keyService, stateProvider);
+  }
+
+  /** {@link ForwarderGeneratorStrategy.generate} */
```
Comment (djsmith85) [inline libs/common/src/tools/generator/username/forwarders/forward-email.ts:30]: Missing documentation and current line likely belongs to the generate method below

File: libs/common/src/tools/generator/username/forwarders/forward-email.ts [hunk 2]
```diff
@@ -1,25 +1,39 @@
...
 
 /** Generates a forwarding address for Forward Email */
-export class ForwardEmailForwarder implements Forwarder {
+export class ForwardEmailForwarder extends ForwarderGeneratorStrategy<
+  ApiOptions & EmailDomainOptions
+> {
   /** Instantiates the forwarder
    *  @param apiService used for ajax requests to the forwarding service
    *  @param i18nService used to look up error strings
    */
```
Comment (djsmith85) [inline libs/common/src/tools/generator/username/forwarders/forward-email.ts:19]: Ctor documentation needs to be updated

File: libs/common/src/tools/generator/username/forwarders/firefox-relay.ts [hunk 3]
```diff
@@ -1,31 +1,48 @@
...
     private i18nService: I18nService,
-  ) {}
+    encryptService: EncryptService,
+    keyService: CryptoService,
+    stateProvider: StateProvider,
+  ) {
+    super(encryptService, keyService, stateProvider);
+  }
+
+  /** {@link ForwarderGeneratorStrategy.generate} */
```
Comment (djsmith85) [inline libs/common/src/tools/generator/username/forwarders/firefox-relay.ts:27]: Missing documentation and current line likely belongs to the generate method below

**Review by djsmith85 — CHANGES_REQUESTED** [review]
Overall changes look good.

Only really nits to update the docs

File: libs/common/src/tools/generator/username/forwarders/fastmail.ts [hunk 4]
```diff
@@ -1,24 +1,36 @@
...
-import { EmailPrefixOptions, Forwarder, ApiOptions } from "../options/forwarder-options";
+import { EmailPrefixOptions, ApiOptions } from "../options/forwarder-options";
 
 /** Generates a forwarding address for Fastmail */
-export class FastmailForwarder implements Forwarder {
+export class FastmailForwarder extends ForwarderGeneratorStrategy<ApiOptions & EmailPrefixOptions> {
   /** Instantiates the forwarder
    *  @param apiService used for ajax requests to the forwarding service
    *  @param i18nService used to look up error strings
    */
```
[the remark this answers is not in the document]
Comment (audrey-jensen) [reply libs/common/src/tools/generator/username/forwarders/fastmail.ts]: Addressed in b63818c2.

File: libs/common/src/tools/generator/username/forwarders/forward-email.ts (hunk 2 above)
Comment (audrey-jensen) [reply libs/common/src/tools/generator/username/forwarders/forward-email.ts]: Addressed in b63818c2.

File: libs/common/src/tools/generator/username/forwarders/simple-login.ts [hunk 5]
```diff
@@ -1,21 +1,36 @@
...
-import { Forwarder, SelfHostedApiOptions } from "../options/forwarder-options";
+import { SelfHostedApiOptions } from "../options/forwarder-options";
 
 /** Generates a forwarding address for Simple Login */
-export class SimpleLoginForwarder implements Forwarder {
+export class SimpleLoginForwarder extends ForwarderGeneratorStrategy<SelfHostedApiOptions> {
   /** Instantiates the forwarder
    *  @param apiService used for ajax requests to the forwarding service
    *  @param i18nService used to look up error strings
    */
```
[the remark this answers is not in the document]
Comment (audrey-jensen) [reply libs/common/src/tools/generator/username/forwarders/simple-login.ts]: Addressed in b63818c2.

File: libs/common/src/tools/generator/username/forwarders/addy-io.ts [hunk 6]
```diff
@@ -1,24 +1,38 @@
...
     private i18nService: I18nService,
-  ) {}
+    encryptService: EncryptService,
+    keyService: CryptoService,
+    stateProvider: StateProvider,
+  ) {
+    super(encryptService, keyService, stateProvider);
+  }
+
+  /** {@link ForwarderGeneratorStrategy.generate} */
```
[the remark this answers is not in the document]
Comment (audrey-jensen) [reply libs/common/src/tools/generator/username/forwarders/addy-io.ts]: Addressed in 98a943a9.

File: libs/common/src/tools/generator/username/forwarders/duck-duck-go.ts [hunk 7]
```diff
@@ -1,21 +1,36 @@
...
     private i18nService: I18nService,
-  ) {}
+    encryptService: EncryptService,
+    keyService: CryptoService,
+    stateProvider: StateProvider,
+  ) {
+    super(encryptService, keyService, stateProvider);
+  }
+
+  /** {@link ForwarderGeneratorStrategy.generate} */
```
[the remark this answers is not in the document]
Comment (audrey-jensen) [reply libs/common/src/tools/generator/username/forwarders/duck-duck-go.ts]: Addressed in 98a943a9.

File: libs/common/src/tools/generator/username/forwarders/fastmail.ts [hunk 8]
```diff
@@ -1,24 +1,36 @@
...
     private i18nService: I18nService,
-  ) {}
+    encryptService: EncryptService,
+    keyService: CryptoService,
+    stateProvider: StateProvider,
+  ) {
+    super(encryptService, keyService, stateProvider);
+  }
+
+  /** {@link ForwarderGeneratorStrategy.generate} */
```
[the remark this answers is not in the document]
Comment (audrey-jensen) [reply libs/common/src/tools/generator/username/forwarders/fastmail.ts]: Addressed in 98a943a9.

File: libs/common/src/tools/generator/username/forwarders/firefox-relay.ts (hunk 3 above)
Comment (audrey-jensen) [reply libs/common/src/tools/generator/username/forwarders/firefox-relay.ts]: Addressed in 98a943a9.

File: libs/common/src/tools/generator/username/forwarders/forward-email.ts (hunk 1 above)
Comment (audrey-jensen) [reply libs/common/src/tools/generator/username/forwarders/forward-email.ts]: Addressed in 98a943a9.

File: libs/common/src/tools/generator/username/forwarders/simple-login.ts [hunk 9]
```diff
@@ -1,21 +1,36 @@
...
     private i18nService: I18nService,
-  ) {}
+    encryptService: EncryptService,
+    keyService: CryptoService,
+    stateProvider: StateProvider,
+  ) {
+    super(encryptService, keyService, stateProvider);
+  }
+
+  /** {@link ForwarderGeneratorStrategy.generate} */
```
[the remark this answers is not in the document]
Comment (audrey-jensen) [reply libs/common/src/tools/generator/username/forwarders/simple-login.ts]: Addressed in 98a943a9.

**Review by djsmith85 — APPROVED** [review]

**Pull request merged**